reasoner/resolution/framework/AnswerCache.java

```diff
@@ -37,42 +37,51 @@
 import java.util.Map;
 import java.util.Optional;
 import java.util.Set;
+import java.util.function.Supplier;
 
 import static com.vaticle.typedb.core.common.exception.ErrorMessage.Internal.ILLEGAL_CAST;
 import static com.vaticle.typedb.core.common.exception.ErrorMessage.Internal.ILLEGAL_STATE;
+import static com.vaticle.typedb.core.common.iterator.Iterators.empty;
 import static com.vaticle.typedb.core.common.iterator.Iterators.iterate;
-import static com.vaticle.typedb.core.common.poller.Pollers.poll;
 
 public abstract class AnswerCache<ANSWER, SUBSUMES> {
 
     protected final List<ANSWER> answers;
     private final Set<ANSWER> answersSet;
     protected boolean reiterateOnAnswerAdded;
     protected boolean requiresReiteration;
-    protected Poller<ANSWER> answerSource;
+    protected FunctionalIterator<ANSWER> answerSource;
     protected boolean complete;
     protected final Map<SUBSUMES, ? extends AnswerCache<?, SUBSUMES>> cacheRegister;
     protected final ConceptMap state;
+    private final Supplier<FunctionalIterator<ANSWER>> answerSourceSupplier;
+    private boolean sourceCleared;
+    private boolean sourceExhausted;
 
-    protected AnswerCache(Map<SUBSUMES, ? extends AnswerCache<?, SUBSUMES>> cacheRegister, ConceptMap state) {
+    protected AnswerCache(Map<SUBSUMES, ? extends AnswerCache<?, SUBSUMES>> cacheRegister, ConceptMap state,
+                          Supplier<FunctionalIterator<ANSWER>> answerSourceSupplier) {
```

Comment:
We only ever use one source, and it's always an iterator, so that lets us simplify. Mainly doing this because we need it. It means that we are able to kill the traversal when we do subsumption, where previously we got the cache to set itself as complete because it was aware of the other caches.